primitives/src/vrf/blsvrf.rs

```diff
@@ -2,26 +2,51 @@
 
 use super::Vrf;
 use crate::{
-    constants::CS_ID_BLS_VRF_NAIVE,
     errors::PrimitivesError,
     signatures::{
         bls::{BLSSignKey, BLSSignature, BLSVerKey},
         BLSSignatureScheme, SignatureScheme,
     },
 };
-use ark_std::rand::{CryptoRng, RngCore};
-use digest::Digest;
+use ark_std::{
+    boxed::Box,
+    rand::{CryptoRng, RngCore},
+    vec::Vec,
+};
+use digest::{Digest, DynDigest};
+use sha2::{Sha256, Sha512};
+
+/// Supported Cipher Suites for BLS VRF.
+#[allow(non_camel_case_types)]
+#[derive(Debug)]
+pub enum BLSVRFCipherSuite {
+    /// using blst library and VRF output from SHA256 hashing
+    VRF_BLS_12_381_SHA256,
+    /// using blst library and VRF output from SHA512 hashing
+    VRF_BLS_12_381_SHA512,
+}
 
 /// BLS VRF scheme.
 /// Optimized for signature size, i.e.: PK in G2 and sig in G1
-pub struct BLSVRFScheme;
+pub struct BLSVRFScheme {
+    hasher: Box<dyn DynDigest>,
+}
 
-impl<H> Vrf<H> for BLSVRFScheme
-where
-    H: Digest,
-{
-    const CS_ID: &'static str = CS_ID_BLS_VRF_NAIVE;
+impl BLSVRFScheme {
+    /// Creates a new BLS VRF instance with the given ciphersuite.
+    pub fn new(cs_id: BLSVRFCipherSuite) -> Self {
+        match cs_id {
+            BLSVRFCipherSuite::VRF_BLS_12_381_SHA256 => Self {
+                hasher: Box::new(Sha256::new()),
+            },
+            BLSVRFCipherSuite::VRF_BLS_12_381_SHA512 => Self {
+                hasher: Box::new(Sha512::new()),
+            },
```

Comment:
I'm slightly concerned about the tradeoff of using trait objects here (i.e. `Box<dyn>`)

I understood it as a way around not having to pass in as a generic param on the trait`VrfScheme<H: Hasher>`, but dynamic dispatch come at a runtime cost ([some reports](https://medium.com/digitalfrontiers/rust-dynamic-dispatching-deep-dive-236a5896e49b) say about 20% slowdown), and our VRF is quite performance sensitive.

imo, the following idea might be better: 
```rust
// trait is not generic, as some VRF instantiation uses PRF instead of Hash to get the output e.g.
trait Vrf {
    fn prove() {}
}

// the concrete impl will be generic
struct BlsVrf<H: Digest>(PhantomData<H>);

impl<H: Digest> Vrf for BlsVrf<H> {
    fn prove() {
        let hasher = H::new();
    }
}
```

Reply:
Your proposed solution was actually my first attempt at this. The dynamic trait is indeed added to avoid the generic `H`. Also it allows for a fine control of which concrete instantiations are actually supported via the `BLSVRFCipherSuite` enum.
The concern about the performance is a good point though. I will run some benchmarks on both approaches.

Reply:
@alxiong Indeed your proposed solution is faster. I've ran some benchmarks [in this branch](https://github.com/EspressoSystems/jellyfish/blob/vrf-bench/bench.md#vrf-benchmarks) and found about 13% speed difference for hashing (no difference for the other operations, as expected).

Taking things into perspective though, we should evaluate whether a 13% speedup on the hashing (which itself is about 1/1000th of the verification time) is worth the convenience that comes with a dynamic hasher and the control that is implied by the use of enums to instantiate only the allowed hashers.

Reply:
> a 13% speedup on the hashing (which itself is about 1/1000th of the verification time) is worth the convenience

good point, I'm on you side now.